### PR TITLE
Add Spring Boot 3 integration modules

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

Spring 6 and Spring Boot 3 have been released.
https://spring.io/blog/2022/11/24/spring-boot-3-0-goes-ga

Modifications:

- New modules are added:
  - spring-boot3-{autoconfigure,starter}
  - spring-boot3-actuator-{autoconfigure,starter}
  - spring-boot3-webflux-{autoconfigure,starter}
  - tomcat10
- Use Java 17 for Spring Boot 3 module, Spring examples and integration tests.
- Use Java 11 for Tomcat 10 module.
- Migrate removed or changed API
  - Use Jakarta EE 9 APIs (`jakarta.*`) instead of EE 8 (`javax.*`).
  - Add `PortUtil` and migrate off `SocketUtils`.
  - Migrate AutoConfiguration in `META-INF/spring.factories` to `o.s.b.a.AutoConfiguration.imports`. https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-2.7-Release-Notes#new-autoconfiguration-annotation
  - Change `management.metrics.export.defaults.enabled` to `management.promethus.metrics.export.enabled` in tests https://github.com/spring-projects/spring-boot/issues/30381
  - Fix compile errors in the Webflux integration
    - `AbstractServerHttpRequest.getMethod()` became non-default method.
  - Migrate relocated `LocalServerPort` and `LocalManagementPort`.
    - `org.springframework.boot.web.server.LocalServerPort` -> `org.springframework.boot.test.web.server.LocalServerPort`

Dependencies:

- Add Spring Boot 3.0.2
- Add hibernate-validator 8.0.0
- Bump Micrometer into 1.10.2
- Add Jakarta dependencies
  - jakarta-inject 2.0.1
  - jakarta-validation 3.0.2
  - jakarta-websocket 2.1.0

Result:

You can now use Spring Boot 3 with Armeria.

### CODE REVIEW

File: spring/boot3-webflux-autoconfigure/src/main/java/com/linecorp/armeria/spring/web/reactive/ArmeriaServerHttpRequest.java
Comment (ikhoon) [inline spring/boot3-webflux-autoconfigure/src/main/java/com/linecorp/armeria/spring/web/reactive/ArmeriaServerHttpRequest.java:118]:
```suggestion
    // This method became a non-default method in Spring 6.
```

File: zookeeper3/src/test/java/com/linecorp/armeria/common/zookeeper/ZooKeeperTestUtilTest.java
Comment (ikhoon) [inline zookeeper3/src/test/java/com/linecorp/armeria/common/zookeeper/ZooKeeperTestUtilTest.java:31]:
```suggestion
        assertThat(ports).doesNotContain(0)
                         .hasSize(5);
```

**Comment by ikhoon** [comment]
Rescheduled the milestone to 1.23.0

**Comment by cj848** [comment]
When can this PR be merged? This is a very much awaited PR.

File: spring/boot3-webflux-autoconfigure/src/main/java/com/linecorp/armeria/spring/web/reactive/ArmeriaServerHttpResponse.java
Comment (minwoox) [inline spring/boot3-webflux-autoconfigure/src/main/java/com/linecorp/armeria/spring/web/reactive/ArmeriaServerHttpResponse.java:55]: So this class now extends to two different classes depending on the module.
How about making it the opposite which means the version-specific implementation extends to  `ArmeriaServerHttpResponse`?

File: spring/boot3-actuator-autoconfigure/build.gradle
Comment (minwoox) [inline spring/boot3-actuator-autoconfigure/build.gradle:44]: Shouldn't we do this separately? `tasks.compileTestJava.dependsOn(generateTestSources)`

File: spring/boot3-webflux-autoconfigure/build.gradle
Comment (minwoox) [inline spring/boot3-webflux-autoconfigure/build.gradle:18]: I'm not sure if we need this or not. Shouldn't the user who wants to use WebSocket include this dependency?

File: zookeeper3/src/test/java/com/linecorp/armeria/common/zookeeper/ZooKeeperTestUtilTest.java
Comment (minwoox) [inline zookeeper3/src/test/java/com/linecorp/armeria/common/zookeeper/ZooKeeperTestUtilTest.java:25]: Flaky?

**Review by minwoox — COMMENTED** [review]
Excellent work! 👍 👍 👍 
Left nits and questions.

**Comment by minwoox** [comment]
> When can this PR be merged? This is a very much awaited PR.

Sorry about the delay. 😓 
I think this is almost done and I hope we could include this in the next release. Please stay tuned. 🙇

File: core/src/main/java/com/linecorp/armeria/internal/common/util/PortUtil.java
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/internal/common/util/PortUtil.java:35]: Oops... 😱

File: core/src/main/java/com/linecorp/armeria/internal/common/util/PortUtil.java
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/internal/common/util/PortUtil.java:29]: Makes sense.

File: gradle/scripts/lib/java.gradle
Comment (ikhoon) [reply gradle/scripts/lib/java.gradle:126]: `javaN` flag and `javaSourceCompatibility` are mutually exclusive. I thought we may give `javaN` higher priority over `javaSourceCompatibility`. Because `javaSourceCompatibility` can be configured globally for all modules but `javaN` flag can be set to each module level.

File: it/spring/webflux-security/src/test/java/com/linecorp/armeria/spring/security/SecureConfiguration.java
Comment (minwoox) [reply it/spring/webflux-security/src/test/java/com/linecorp/armeria/spring/security/SecureConfiguration.java:29]: Oops... 😢

File: gradle/scripts/lib/java.gradle
Comment (minwoox) [reply gradle/scripts/lib/java.gradle:126]: That makes sense. Thanks for the explanation. 🙇

File: core/src/test/java/com/linecorp/armeria/internal/common/util/PortUtilTest.java
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/internal/common/util/PortUtilTest.java:34]: This can always fail. If really need to use this instead of binding at port 0, we should enforce retries.

File: spring/boot3-autoconfigure/src/test/resources/config/application-autoConfTest.yml
Comment (trustin) [inline spring/boot3-autoconfigure/src/test/resources/config/application-autoConfTest.yml:7]: Maybe remove this comment from the .yml files for Spring Boot 3? This is useful for migration but these files were done with migration 😉

File: zookeeper3/src/test/java/com/linecorp/armeria/common/zookeeper/ZooKeeperTestUtil.java
Comment (trustin) [inline zookeeper3/src/test/java/com/linecorp/armeria/common/zookeeper/ZooKeeperTestUtil.java:45]: Could be package-private + `@VisibleForTesting`?

File: reactor3/src/test/java/com/linecorp/armeria/common/reactor3/ContextAwareFluxTest.java
Comment (ikhoon) [reply reactor3/src/test/java/com/linecorp/armeria/common/reactor3/ContextAwareFluxTest.java:393]: Yes, they are. 😅

File: spring/boot3-webflux-autoconfigure/src/main/java/com/linecorp/armeria/spring/web/reactive/ArmeriaServerHttpResponse.java
Comment (ikhoon) [reply spring/boot3-webflux-autoconfigure/src/main/java/com/linecorp/armeria/spring/web/reactive/ArmeriaServerHttpResponse.java:55]: It may be possible. But as you see, the name is pretty verbose. so I didn't want to use it in other places and hide it in the super class.

File: spring/boot3-actuator-autoconfigure/build.gradle
Comment (ikhoon) [reply spring/boot3-actuator-autoconfigure/build.gradle:44]: `generateSources` depends on `generateTestSources` and `compileTestJava` depends on `compileJava`. So `generateTestSources` will be invoked before `compileTestJava`

File: spring/boot3-webflux-autoconfigure/build.gradle
Comment (ikhoon) [reply spring/boot3-webflux-autoconfigure/build.gradle:18]: https://github.com/spring-projects/spring-framework/blob/main/spring-webflux/src/main/java/org/springframework/web/reactive/socket/server/support/HandshakeWebSocketService.java#L272-L291
WebFlux in Spring Boot 3 uses Jakarta Websocket as a fallback if no custom WebSocket `RequestUpgradeStrategy`. The explanation was not enough. Let me update the comment.

File: core/src/test/java/com/linecorp/armeria/internal/common/util/PortUtilTest.java
Comment (ikhoon) [reply core/src/test/java/com/linecorp/armeria/internal/common/util/PortUtilTest.java:34]: Agreed. Let me mark it as `@FlakyTest`.

File: zookeeper3/src/test/java/com/linecorp/armeria/common/zookeeper/ZooKeeperTestUtil.java
Comment (ikhoon) [reply zookeeper3/src/test/java/com/linecorp/armeria/common/zookeeper/ZooKeeperTestUtil.java:45]: As it is already in test, let me remove `public` only.

File: tomcat9/build.gradle
Comment (jrhee17) [inline tomcat9/build.gradle:22]: Question) Is there no need for `compileTestJava` to depend on `generateSources` (or `compileJava`)?
I realize that this is just a copy of `tomcat8`, but this seemed odd 🤔

**Review by jrhee17 — COMMENTED** [review]
Left only nits and questions 😄    Thanks for the handling this issue @ikhoon 🙇

**Review by trustin — APPROVED** [review]
LGTM once the conflicts are resolved. Thanks for adding this long-awaited module!

File: gradle.properties
Comment (ikhoon) [reply gradle.properties:28]: Good point. It needs reverting. I didn't know `no_aggregation` disabled Jacoco reporting.

File: spring/boot3-autoconfigure/src/test/resources/config/application-autoConfTest.yml
Comment (ikhoon) [reply spring/boot3-autoconfigure/src/test/resources/config/application-autoConfTest.yml:7]: Yes, I put both properties to share this configuration file. We can check out the change properties in
https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-3.0.0-Configuration-Changelog
```diff
- management.metrics.export.prometheus.enabled
+ management.prometheus.metrics.export.enabled
```

File: spring/boot3-webflux-autoconfigure/src/main/java/com/linecorp/armeria/spring/web/reactive/AbstractServerHttpResponse.java
Comment (ikhoon) [reply spring/boot3-webflux-autoconfigure/src/main/java/com/linecorp/armeria/spring/web/reactive/AbstractServerHttpResponse.java:182]: >  If I understand correctly this also affects upstream. 

It might be a problem only on our side. If we didn't recover the `CancelledSubscriptionException` exception,
https://github.com/line/armeria/blob/master/spring/boot3-webflux-autoconfigure/src/main/java/com/linecorp/armeria/spring/web/reactive/ArmeriaServerHttpResponse.java#L93-L95
the buffer would be released by 
https://github.com/line/armeria/blob/96ee8e871b2834ed765e83d3fd317072fce346e4/spring/boot3-webflux-autoconfigure/src/main/java/com/linecorp/armeria/spring/web/reactive/AbstractServerHttpResponse.java#L187

Anyway, it would be worth discussing it with upstream maintainers. Let me create an issue for it.

File: spring/boot3-webflux-autoconfigure/build.gradle
Comment (minwoox) [reply spring/boot3-webflux-autoconfigure/build.gradle:21]: Thanks a lot for the explanation. 🙇

**Review by minwoox — APPROVED** [review]
Great job, @ikhoon!
Thanks a lot! 😄

File: spring/boot3-autoconfigure/src/test/resources/config/application-settings.yml
Comment (ikhoon) [reply spring/boot3-autoconfigure/src/test/resources/config/application-settings.yml]: It seems my mistake. They were removed.

**Review by jrhee17 — APPROVED** [review]
Looks really nice! Thanks @ikhoon ! 👍 🚀 🙇

**Comment by ikhoon** [comment]
Thanks for the review. I missed pushing some changes that should be checked in after the final review. [`224489a` (#4574)](https://github.com/line/armeria/pull/4574/commits/224489a196623dca14b6418885e852c35eded36c) The changes are trivial ones related to Jacoco and Javadoc.

**Review by trustin — APPROVED** [review]
Still LGTM 🚀

**Pull request merged**
